### PR TITLE
RAB-935: Add UpsertUser ServiceAPI 

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**

In this PR, we added the a new PublicAPI to create a user. For now we only need it for job user type

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/UserManagement/API/User/UpsertUserCommand.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Akeneo\UserManagement\API\User;
```
Comment (ghoussard) [inline src/Akeneo/UserManagement/API/User/UpsertUserCommand.php:10]: Should be `ServiceAPI`

**Review by olivierpontier — APPROVED** [review]

File: src/Akeneo/UserManagement/API/User/UpsertUserCommand.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
```
Comment (MadeWilson) [inline src/Akeneo/UserManagement/API/User/UpsertUserCommand.php:1]: should'nt API directory be under infrastructure and named ServiceApi

File: src/Akeneo/UserManagement/API/User/UpsertUserCommand.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Akeneo\UserManagement\API\User;
+
+final class UpsertUserCommand
+{
+    public function __construct(
```
Comment (ghoussard) [inline src/Akeneo/UserManagement/API/User/UpsertUserCommand.php:14]: We should say that $rolesCodes and $groupCodes are string array with a PHPDoc

File: src/Akeneo/UserManagement/API/User/UpsertUserCommand.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        public string $username,
+        public string $password,
+        public string $email,
+        public string $type,
+        public string $firstName,
+        public string $lastName,
+        public array $roleCodes,
+        public array $groupCodes = [],
+    ) {
+    }
```
Comment (ghoussard) [inline src/Akeneo/UserManagement/API/User/UpsertUserCommand.php:24]: Maybe we can do some assertions here like asserting not blank $username or $rolesCodes is an array of string

File: src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            if ($user->getType() !== $upsertUserCommand->type) {
+                throw new InvalidPropertyException(
+                    'type',
+                    $upsertUserCommand->type,
+                    null,
+                    "You could not change the user type"
+                );
+            }
+
+            $userPayload = [
```
Comment (ghoussard) [inline src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php:56]: So we allow to update the username ?

File: src/Akeneo/UserManagement/Component/Model/User.php [hunk 6]
```diff
@@ -553,7 +554,7 @@ public function unsetImageFile()
      */
     public function setEnabled($enabled)
     {
-        $this->enabled = (boolean) $enabled;
+        $this->enabled = (bool) $enabled;
```
Comment (ghoussard) [inline src/Akeneo/UserManagement/Component/Model/User.php:557]: 🍑

**Review by ghoussard — APPROVED** [review]

File: src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php (hunk 5 above)
Comment (MadeWilson) [inline src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php:56]: I would speak of userData, payload is something that is shared across multiple service in a pipeline execution for me.

File: src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            throw new ViolationsException($constraintViolations);
+        }
+
+        $this->userSaver->save($user);
+    }
+
+    private function defineType(UserInterface $user, string $type): void
+    {
+        switch ($type) {
+            case 'user':
```
Comment (MadeWilson) [inline src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php:94]: don't we have those value in a const some where ?

**Review by MadeWilson — APPROVED** [review]

File: src/Akeneo/UserManagement/Application/Storage/FindUserGroups.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+interface FindUserGroups
+{
+    public const DEFAULT_LIMIT = 25;
+
+    /** @return Group[] */
+    public function __invoke(
+        ?string $search = null,
+        ?int $searchAfterId = null,
+        int $limit = self::DEFAULT_LIMIT
```
Comment (grena) [inline src/Akeneo/UserManagement/Application/Storage/FindUserGroups.php:15]: I thought we were adopting the same design as we did with Family ServiceAPI; a `FindUserSearch` object. This way later if we want to search by `type` or `email`, we can update this object and handle it in the handler without changing its signature

File: src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            $this->defineType($user, $upsertUserCommand->type);
+        }
+
+        try {
+            if ($user->getType() !== $upsertUserCommand->type) {
+                throw new InvalidPropertyException(
+                    'type',
+                    $upsertUserCommand->type,
+                    null,
+                    "You could not change the user type"
```
Comment (grena) [inline src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php:52]:
```suggestion
                    "You cannot change the user type"
```

File: src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php (hunk 7 above)
Comment (grena) [reply src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php:94]: Yup, maybe we could use `User::TYPE_API`, `User::TYPE_USER`, and make them `public` if they aren't.

File: src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            case 'api':
+                $user->defineAsApiUser();
+
+                return;
+            case 'job':
+                $user->defineAsJobUser();
+
+                return;
+            default:
+                throw new \InvalidArgumentException(sprintf('Invalid type "%s"', $type));
```
Comment (grena) [inline src/Akeneo/UserManagement/Application/Handler/User/UpsertUserHandler.php:105]:
```suggestion
                throw new \InvalidArgumentException(sprintf('Invalid user type "%s"', $type));
```

File: src/Akeneo/UserManagement/Domain/Model/Group.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        return self::DEFAULT_TYPE === $this->type;
+    }
+
+    public function getDefaultPermissions(): array
+    {
+        return $this->defaultPermissions;
+    }
+
+    public static function createFromDatabase(array $data): self
```
Comment (grena) [inline src/Akeneo/UserManagement/Domain/Model/Group.php:42]:
```suggestion
    public static function fromNormalized(array $data): self
```

File: src/Akeneo/UserManagement/API/User/UpsertUserCommand.php (hunk 2 above)
Comment (MadeWilson) [reply src/Akeneo/UserManagement/API/User/UpsertUserCommand.php:1]: @ValentinMumble and @olivierpontier talk about this during the last design meeting I'm doing it in the UserGroup PR

**Review by ValentinMumble — APPROVED** [review]

**Pull request merged**